app/src/behaviors/behavior_tap_dance.c

```diff
@@ -0,0 +1,340 @@
+        // Tap dance is currently pressed, prepare for release of binding
+        // before actual behavior release
+        if (tap_dance->is_pressed)
+            return;
+        // Tap dance is not currently pressed. Release the tap dance.
+        LOG_DBG("Tap dance releases immediately! Counter reached: %d", tap_dance->counter);
+        release_tap_dance_behavior(tap_dance, tap_dance->release_at);
+        return;
+    }
+}
+
+static const struct behavior_driver_api behavior_tap_dance_driver_api = {
+    .binding_pressed = on_tap_dance_binding_pressed,
+    .binding_released = on_tap_dance_binding_released,
+};
+
+// tap_dance_position_state_changed_listener
+
+// Declared here, defined below. Determines what happens when another
+// keycode is pressed other than any active tap-dances. When any
+// tap-dance is interrupted, they will resolve before the interruption
+// keycode and wait to release when their binding is released
+static int tap_dance_position_state_changed_listener(const zmk_event_t *eh);
+
+ZMK_LISTENER(behavior_tap_dance, tap_dance_position_state_changed_listener);
+ZMK_SUBSCRIPTION(behavior_tap_dance, zmk_position_state_changed);
+
+static int tap_dance_position_state_changed_listener(const zmk_event_t *eh) {
+    // A keypress with a different position than the last tap dance has been recorded.
+    struct zmk_position_state_changed *ev = as_zmk_position_state_changed(eh);
+    if (ev == NULL) {
+        return ZMK_EV_EVENT_BUBBLE;
+    }
+    for (int i = 0; i < ZMK_BHV_TAP_DANCE_MAX_HELD; i++) {
+        struct active_tap_dance *tap_dance = &active_tap_dances[i];
+        if (tap_dance->position == ZMK_BHV_TAP_DANCE_POSITION_FREE) {
+            continue;
+        }
+        if (tap_dance->position != ev->position && tap_dance->timer_started) {
+            // Press tap dance immediately before new keypress.
+            stop_timer(tap_dance);
+            LOG_DBG("Tap dance interrupted, position changed. %d %d", tap_dance->position,
+                    ev->position);
+            if (!tap_dance->tap_dance_decided) {
+                tap_dance->tap_dance_decided = true;
+                // Ensure tap-dance resolves before next keycode
+                ZMK_EVENT_RAISE(press_tap_dance_behavior(tap_dance, ev->timestamp));
+                ZMK_EVENT_RAISE(eh);
+                return ZMK_EV_EVENT_CAPTURED;
+            }
+            // Tap dance will release on key release, allowing for decided tap-dances
+            // to be held, like those using mod keys
+        }
+    }
+    return ZMK_EV_EVENT_BUBBLE;
+}
+
+static int behavior_tap_dance_init(const struct device *dev) {
+    static bool init_first_run = true;
+    if (init_first_run) {
+        for (int i = 0; i < ZMK_BHV_TAP_DANCE_MAX_HELD; i++) {
+            k_delayed_work_init(&active_tap_dances[i].release_timer,
+                                behavior_tap_dance_timer_handler);
+            active_tap_dances[i].position = ZMK_BHV_TAP_DANCE_POSITION_FREE;
```

Comment:
call `clear_tap_dance` instead of setting the position here manually. That way tapdances are always cleared in the same way.